grid-client/calculator/calculate.go

```diff
@@ -143,6 +152,314 @@ func calculateCU(cru, mru int64) float64 {
+
+	contractMonthlyCostUSD, err := c.calculateContractCost(contract)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to calculate contract cost")
+	}
+
+	contractMonthlyCostTFT, err := c.USDtoTFT(contractMonthlyCostUSD)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert contract cost to TFT")
+	}
+	secondsPerMonthBig := big.NewFloat(30 * 24 * 60 * 60) // 30 days * 24 hours * 60 minutes * 60 seconds
+	contractCostPerSecond := new(big.Float).Quo(contractMonthlyCostTFT, secondsPerMonthBig)
+	totalPeriodCost := new(big.Float).Mul(contractCostPerSecond, big.NewFloat(totalPeriodSeconds))
+	return totalPeriodCost, nil
+
+}
+
+// Calculates the cost of a contract per month in USD.
+func (c *Calculator) calculateContractCost(contract *substrate.Contract) (float64, error) {
+	if contract.ContractType.IsNameContract {
+		return c.calculateUniqueNameCost()
+	}
+
+	nodeID, err := getNodeID(contract)
+	if err != nil {
+		return 0, err
+	}
+
+	node, err := c.substrateConn.GetNode(nodeID)
+	if err != nil {
+		return 0, err
+	}
+
+	var nodeRentContract uint64
+
+	nodeRentContract, err = c.substrateConn.GetNodeRentContract(nodeID)
+	if err != nil && !errors.Is(err, substrate.ErrNotFound) {
+		return 0, err
+	}
+
+	if contract.ContractType.IsNodeContract {
+		return c.calculateNodeContractCost(contract, node.Certification.IsCertified, nodeRentContract > 0)
+	}
+
+	if contract.ContractType.IsRentContract {
+
+		return c.calculateRentCost(contract, *node)
+	}
+	return 0, nil
+}
+
+// Calculates the cost of a unique name per month in USD.
+func (c *Calculator) calculateUniqueNameCost() (float64, error) {
+	//TODO should we apply stacking discount?
+	pricingPolicy, err := c.substrateConn.GetPricingPolicy(defaultPricingPolicyID)
+	if err != nil {
+		return 0, err
+	}
+	// cost in unit-USD
+	monthlyCost := float64(pricingPolicy.UniqueName.Value) * 24 * 30
+	return float64(monthlyCost) / UnitFactor, nil
+}
+
+// Calculates the cost of a node contract per month in USD.
+//
+// There are two cases for node contract cost:
+//  1. Node contract on shared node: the cost of the node (shared)
+//  2. Node contract on rented node: the cost of the IPV4 only if the contact includes ipv4, else it will return zero.
+func (c *Calculator) calculateNodeContractCost(contract *substrate.Contract, onCertifiedNode, isOnRentedNode bool) (float64, error) {
+	if !contract.ContractType.IsNodeContract {
+		return 0, fmt.Errorf("contract ID %d is not a node contract", contract.ContractID)
+	}
+	publicIPsCount := contract.ContractType.NodeContract.PublicIPsCount
+
+	isCertified := node.Certification.IsCertified
+
+	// Node contract on rented node
+	if isOnRentedNode {
+		if publicIPsCount > 0 {
+			cost, err := c.calculateIPV4()
+			if err != nil {
+				return 0, err
+			}
+			totalCost := cost * float64(publicIPsCount)
+
+			//TODO should we apply stacking discount?
+
+			if onCertifiedNode {
+				totalCost *= 1.25
+			}
+			return totalCost, nil
+		}
+		return 0, nil
```

Comment:
you can check if publicIPsCount is zero and return immediately having the logic outside of the inner if for less indentation 
```suggestion
		if publicIPsCount == 0 {
			return 0, nil
		}
		cost, err := c.calculateIPV4()
		if err != nil {
			return 0, err
		}
		totalCost := cost * float64(publicIPsCount)

		// TODO should we apply stacking discount?

		if onCertifiedNode {
			totalCost *= 1.25
		}
		return totalCost, nil

```